actors/builtin/miner/sector_map.go

```diff
@@ -0,0 +1,151 @@
+package miner
+
+import (
+	"math"
+	"sort"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-bitfield"
+)
+
+// Maps deadlines to partition maps.
+type DeadlineSectorMap map[uint64]PartitionSectorMap
+
+// Maps partitions to sector bitfields.
+type PartitionSectorMap map[uint64]*bitfield.BitField
+
+// Check validates all bitfields and counts the number of partitions & sectors
+// contained within the map, and returns an error if they exceed the given
+// maximums.
+func (dm DeadlineSectorMap) Check(maxPartitions, maxSectors uint64) error {
+	partitionCount, sectorCount, err := dm.Count()
+	if err != nil {
+		return xerrors.Errorf("failed to count sectors: %w", err)
+	}
+	if partitionCount > maxPartitions {
+		return xerrors.Errorf("too many partitions %d, max %d", partitionCount, maxPartitions)
+	}
+
+	if sectorCount > maxSectors {
+		return xerrors.Errorf("too many sectors %d, max %d", sectorCount, maxSectors)
+	}
+
+	return nil
+}
+
+// Count counts the number of partitions & sectors within the map.
+func (dm DeadlineSectorMap) Count() (partitions, sectors uint64, err error) {
+	for dlIdx, pm := range dm { //nolint:nomaprange
+		partCount, sectorCount, err := pm.Count()
+		if err != nil {
+			return 0, 0, xerrors.Errorf("when counting deadline %d: %w", dlIdx, err)
+		}
+		if partCount > math.MaxUint64-partitions {
+			return 0, 0, xerrors.Errorf("uint64 overflow when counting partitions")
+		}
+
+		if sectorCount > math.MaxUint64-sectors {
+			return 0, 0, xerrors.Errorf("uint64 overflow when counting sectors")
+		}
+		sectors += sectorCount
+		partitions += partCount
+	}
+	return partitions, sectors, nil
+}
+
+// Add records the given sector bitfield at the given deadline/partition index.
+func (dm DeadlineSectorMap) Add(dlIdx, partIdx uint64, sectorNos *bitfield.BitField) error {
+	if dlIdx >= WPoStPeriodDeadlines {
+		return xerrors.Errorf("invalid deadline %d", dlIdx)
+	}
+	dl, ok := dm[dlIdx]
+	if !ok {
+		dl = make(PartitionSectorMap, 1)
```

Comment:
The capacity of `1` here isn't informed by knowledge of the eventual size, and may lead to more allocations (but only sometimes less memory). It's probably better to just leave it unspecified.

Reply:
How so? I know I'm going to add _at least_ one element in `dl.Add`, that's why I'm pre-allocating for it.

Looking at the source code, it looks like go always allocates at least one bucket (i.e., if the hint is 0-8, one bucket is allocated regardless).

Reply:
You're probably more familiar with internals (tho they might change). But specifying 1 could cause the allocation of fewer buckets than specifying nothing, which would then cause a re-hashing later when it expands.

Reply:
Well, there are two possible assumptions:

1. Go will always allocate enough space for at least one element.
2. Providing an allocation hint will never cause go to allocate less than it would otherwise.

Either way, we're making an assumption about how go works (and will work) internally. In this case, I'm making the second assumption. I had actually assumed that maps with no hint started at size 0 (which isn't the case), hence the hint of 1.

Reply:
Note: I'm happy to change this, I just feel like it's clearer to tell the compiler "please make room for at least one element" when we know we'll need that space.

Reply:
I had the opposite assumption for 2, but I'm happy to leave it as is.

Reply:
Eh, you're right. It's better to avoid specifying things we don't actually know: https://github.com/filecoin-project/specs-actors/pull/898.